### ISSUE TITLE
Cognito token expired when it shouldn't

### ISSUE THREAD

**Issue by nourahassan** [aws-amplify/aws-sdk-ios#1376]
We are using AWSMobile on iOS with cognito setup. The refresh token expiration is set to 10 years but users are still getting token expiration when trying to fetch user attributes.

Here's the code:

```
AWSMobileClient.sharedInstance().getUserAttributes{ userAttributes, error in
       if let error = error {
              dLog("Couldn't get user attributes \(error)")
       }
}
```

And in log I see this:

`Couldn't get user attributes notAuthorized(message: "Access Token has expired")`

The received error is `AWSMobileClientError.notAuthorized`.

Can you give explanation why we still get this? and how to fix?

**Environment:**
 - SDK Version: 2.9.3
 - Dependency Manager: Carthage
 - Swift Version : 4.2

It happened on different devices.

Not sure if this is related but I can see some timeout exception from AppSync before getting the not authorized exception (the AppSync version is 2.10.3).

```
11:30:39.777   applicationDidEnterBackground..
11:31:59.075   Did not receive a successful HTTP code. Error: Error Domain=NSURLErrorDomain Code=-1001 "The request timed out."
11:31:59.61   applicationWillEnterForeground..
11:31:59.631   Did not receive a successful HTTP code. Error: Error Domain=NSURLErrorDomain Code=-1003 "A server with the specified hostname could not be found." 
11:32:00.19   applicationDidBecomeActive..

11:55:43.737   On background fetch 
11:55:43.775   Did not receive a successful HTTP code. Error: Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." 
11:55:46.483   Couldn't get user attributes notAuthorized(message: "Access Token has expired")
```

**Comment by rohandubal** [comment]
Hello @nourahassan 

I think there are 2 issues here which are causing confusion. 

The 10 years validity which you have is for the `Refresh Token` and not the `Access Token`. The `Refresh Token` is used to fetch a new `AccessToken` when it expires. `AWSMobileClient` internally refreshes the `AccessToken` when it expires(every hour). But what seems to happen is the refresh call is not going though due to network issues. I believe that if you retry the request once the network is more stable the `getUserAttributes` will go through. Could you please confirm that?

The AppSync issue is completely unrelated and should not impact token expiry. 

Thanks,
Rohan

**Comment by nourahassan** [comment]
So you suggest that when I got token expired exception, I check connectivity? when connectivity is back I will get user attributes without any issue?

Shouldn't I get exception about network timeout or there is no connection instead of token expire?

Not sure how can I test network timeout to confirm your assumption?

**Comment by rohandubal** [comment]
Hello @nourahassan 

I looked deeper into this issue to identify what could possibly cause this.

Regarding the network error; I checked the code and can confirm that the error is propagated back to the developer callback with NSURLErrorDomain and code 1009. So you can look for that code and retry the request.

The specific error of notAuthorized is given by the SDK when the `idToken` is expired. It could also happen when the refresh token is expired. But in the cases that refresh token is expired, we give a notification that an end user sign in is needed. Since you are not receiving that, this issue points to idToken being invalid/ not having enough permissions.

One of the possible reasons this could happen is due to a clock skew on the device. If the device time is slower than actual time (beyond 5 minutes), the SDK will return the current token it has thinking it is still valid. Once the SDK hands the token to the calling entity, the service check of validity of token would fail causing the notAuthorized error which you are seeing.

- Is there a way you can identify if the device seeing this error has any clock skew?
- Does this happen too frequently or only happens to set of devices?

Thanks,
Rohan

**Comment by nourahassan** [comment]
I never get `NSURLErrorDomain and code 1009`, but after reviewing logs for last month I can see that sometimes we got token expired then on the second call we got attributes successfully. 
Can the getUserAttributes not refresh token but other APIs did (like access dynamo or lambda), so when other API refreshes the token the getUserAttributes work on the second time.

We are testing with around 20 devices, on last month 9 of them got the exception at least once. It happened maximum 3 times for the same device. Most of cases with the connection timeout exception before it. 

As workaround now I am saving credentials and do auto login when get the token expired exception.

I am also asking about the clock but I think all ppl select set automatically!

**Comment by rohandubal** [comment]
Hello @nourahassan 

The difference between `getUserAttributes` and dynamodb/ lambda API calls is that `getUserAttributes` uses the JWT access token issued by Cognito User Pool service whereas dynamodb/ lambda use AWS Credentials issued by Cognito Identity service.

The API refresh logic for both are similar. When the credentials are about to expire, they are refreshed by using refresh token / provider tokens. 

Retrying the call would be the right strategy but I think there is an important question to be answered before that -- Do you receive a notification from MobileClient which indicates the status of the user as `signedOutUserPoolsTokenInvalid`? 

In the test devices which you have, have you verified if all the devices have the right time set on them?

**Comment by nourahassan** [comment]
No I didn't get `signedOutUserPoolsTokenInvalid` when I got the error from `getUserAttributes`. I got it when user signed in to another device.

**Comment by rohandubal** [comment]
If you are not receiving that notification, it would mean that the SDK is thinking that the token is valid at the given time. Have you tried retrying the request? Does the request succeed on the retry?

**Comment by nourahassan** [comment]
No I sign the user out then re-login with saved credentials. Btw I didn't get this error last two weeks! I updated the sdk 10 days ago, not sure if something changed.

**Comment by bolak2** [comment]
I am still getting the same error, the S3transfer utility exits with an error: (com.amazonaws.AWSCognitoIdentityErrorDomain error 8

**Comment by nourahassan** [comment]
@rohandubal Good news! I am not seeing errors about couldn't get user attributes because of **token revoked**.

Now I am getting the error because of reasonable reasons related to connection, and attributes are fetched later successfully.

Here's list of errors I saw last couple days:

* The request timed out.
* Software caused connection abort.
* The Internet connection appears to be offline.
* A data connection is not currently allowed.

I think this is fixed now. Thanks!

**Comment by palpatim** [comment]
Since @nourahassan reports their issue appears to be fixed, and @bolak2 has opened #1519 for their issue, I'm closing this one so we don't fragment the conversation.
